aria/src/lib.rs

```diff
@@ -0,0 +1,412 @@
+//! Pure Rust implementation of the [ARIA] block cipher ([RFC 5794]).
+//!
+//! # ⚠️ Security Warning: Hazmat!
+//!
+//! This crate implements only the low-level block cipher function, and is intended
+//! for use for implementing higher-level constructions *only*. It is NOT
+//! intended for direct use in applications.
+//!
+//! USE AT YOUR OWN RISK!
+//!
+//! # Examples
+//! ```
+//! use aria::cipher::generic_array::GenericArray;
+//! use aria::cipher::{Key, Block, BlockEncrypt, BlockDecrypt, KeyInit};
+//! use aria::Aria128;
+//!
+//! let key = GenericArray::from([0u8; 16]);
+//! let mut block = GenericArray::from([0u8; 16]);
+//! // Initialize cipher
+//! let cipher = Aria128::new(&key);
+//!
+//! let block_copy = block.clone();
+//! // Encrypt block in-place
+//! cipher.encrypt_block(&mut block);
+//! // And decrypt it back
+//! cipher.decrypt_block(&mut block);
+//! assert_eq!(block, block_copy);
+//! ```
+//!
+//! [ARIA]: https://en.wikipedia.org/wiki/ARIA_(cipher)
+//! [RFC 5794]: https://tools.ietf.org/html/rfc5794
+
+#![no_std]
+#![doc(
+    html_logo_url = "https://raw.githubusercontent.com/RustCrypto/media/26acc39f/logo.svg",
+    html_favicon_url = "https://raw.githubusercontent.com/RustCrypto/media/26acc39f/logo.svg"
+)]
+#![deny(unsafe_code)]
+#![cfg_attr(docsrs, feature(doc_cfg))]
+#![warn(missing_docs, rust_2018_idioms)]
+
+mod consts;
+
+pub use cipher;
+
+use cipher::{
+    consts::{U16, U24, U32},
+    AlgorithmName, BlockCipher, Key, KeyInit, KeySizeUser,
+};
+use core::fmt;
+
+#[cfg(feature = "zeroize")]
+use cipher::zeroize::{Zeroize, ZeroizeOnDrop};
+
+use crate::consts::{C1, C2, C3, SB1, SB2, SB3, SB4};
+
+macro_rules! diffuse {
+    ($x0: expr, $x1: expr, $x2: expr, $x3: expr,
+     $x4: expr, $x5: expr, $x6: expr, $x7: expr,
+     $x8: expr, $x9: expr, $x10: expr, $x11: expr,
+     $x12: expr, $x13: expr, $x14: expr, $x15: expr,) => {
+        let x = [
+            0x0000_0001_0100_0100_0101_0000_0001_0100 * $x0 as u128,
+            0x0000_0100_0001_0001_0101_0000_0100_0001 * $x1 as u128,
```

Comment:
You can replace this whole macro with the following function:
```rust
#[inline(always)]

pub fn diffuse(x: [u8; 16]) -> u128 {
    DIFFUSE_CONSTS.iter().zip(x).map(|(a, b)| a * b as u128).fold(0, |a, v| a ^ v)
}
```
Compiler is [smart enough](https://rust.godbolt.org/z/a5ceG6heq) to unroll everything.